### ISSUE TITLE
Webpack confused by `React` module name

### ISSUE THREAD

**Issue by pkamenarsky** [purescript-contrib/purescript-react#72]
Whenever there's a `react` dependency in `package.json`, Webpack gets confused by the fact that both `react` and `purescript-react`'s `React.js` need to be placed in the same `output` directory.

Thus when calling `var React = require('react');` from `React.js` it refers to itself (instead of the real `react`) and crashes with something like `React.createClass is not a function`.

Uncommenting that line fixes the issue. I'm not sure what the best way to deal with this would be.

**Comment by ethul** [comment]
One solution is to set the --require-path option to ../ when running psc.
This is the approach used by the purs-loader and purescript-webpack-plugin.

**Comment by jasonzoladz** [comment]
I've had a fairly smooth workflow with react using purs-loader and purescript-webpack-plugin (when I need to see the result in the browser).  It's pretty great.  

Sometimes webpack will gobble up the real compiler errors, however.  In which case, I drop the extension from `entry.js` and build with pulp until I sort the error.

**Comment by pkamenarsky** [comment]
Is there a workaround for `pulp server`?

**Comment by ethul** [comment]
I haven't really tried out `pulp server`, but I can take a look. Maybe @hdgarrood would have some ideas on this. I think if you're able to pass in a `--require-path` value, my then this should do the trick.

**Comment by pkamenarsky** [comment]
That would be great - tell me if you need anything! `--require-path` doesn't seem to work...

**Comment by hdgarrood** [comment]
`pulp server` uses `purs-loader` under the hood, which I'm fairly sure should be supplying a `--require-path` value of `../` by itself. Admittedly it's a rather old version of `purs-loader`.

You probably need to clear your output directory (`rm -r output`) in between building with `pulp build` and `pulp server`, because of purescript/purescript#1836 too. Does that help?

**Comment by pkamenarsky** [comment]
I tried this already (clearing `output` before invoking `pulp server`), doesn't work unfortunately...

**Comment by pkamenarsky** [comment]
Btw, I should add that [shimming](https://github.com/paf31/purescript-thermite/blob/master/html/index.html#L16) `require` doesn't help either - apparently webpack intercepts all calls to `require`.

**Comment by ethul** [comment]
Thanks for the info @hdgarrood 

@pkamenarsky It looks like latest `pulp` is using the `purs-loader@0.4.x`. Can you please indicate your `pulp` version and `purs-loader` version? If you can confirm you are using `purs-loader@0.4.x`, then the require path should be hardcoded to '../'.

Another possibility is that your webpack config might need to be updated. Can you please provide your config or at least your `modulesDirectories` value? You can't include `output` (or where-ever your compiled PuresScript output goes) in your `modulesDirectories`.

Consider the following config:

```javascript
var path = require('path');

var srcs = ['src[]=bower_components/purescript-*/src/**/*.purs', 'src[]=src/**/*.purs'];

var ffis = ['ffi[]=bower_components/purescript-*/src/**/*.js', 'ffi[]=src/**/*FFI.js'];

var output = 'output';

var modulesDirectories = [
  'node_modules',
  'bower_components'
];

var config
  = { entry: './src/entry'
    , output: { path: __dirname
              , pathinfo: true
              , filename: 'bundle.js'
              }
    , module: { loaders: [ { test: /\.purs$/
                           , loader: 'purs-loader?output=' + output + '&' + srcs.concat(ffis).join('&')
                           } ] }
    , resolve: { modulesDirectories: modulesDirectories
               , extensions: ['', '.js']
               }
    , resolveLoader: { root: path.join(__dirname, 'node_modules') }
    }
    ;

module.exports = config;
```

**Comment by pkamenarsky** [comment]
phil$ pulp --version
    Pulp version 8.1.0
    psc version 0.8.0.0 using /usr/local/share/npm/bin/psc

I'm not sure how to find out the version of `purs-loader`, is it a command line utility? If it is, I don't have it.

Also, I don't have a webpack config - I just invoke `pulp server` in my project folder, which looks like this:

    bower.json
    bower_components
    dist
    node_modules
    output
    package.json
    src
    test

Btw, `pulp server` worked perfectly fine before updating to the newest `thermite`, which includes the newest `react` and `react-dom`. I guess in older versions of `react` there was no `require` in `React.js`, which is why it worked.

**Comment by ethul** [comment]
Thanks for the info. To check the purs-loader version you have to look at the `package.json` of the `purs-loader` that is installed in the `node_modules` of `pulp`. Also, there should be a webpack config around. Perhaps it is a hidden file in your directory?

**Comment by hdgarrood** [comment]
There won't necessarily be a webpack config, Pulp bundles a default one so that you don't have to specify one. See: https://github.com/bodil/pulp/blob/v8.1.0/src/Pulp/Server.purs#L73-L106

**Comment by pkamenarsky** [comment]
Ok, found `purs-loader`:

     "purs-loader": "^0.4.0",

And yes, I don't have a webpack config file. @hdgarrood beat me to it, but while debugging the issue I also came across the default webpack config in pulp.

**Comment by ethul** [comment]
Ah. Okay. Thanks @hdgarrood 

I believe the issue is:
https://github.com/bodil/pulp/blob/f98fbbc1e4da22b76a72d09349d876b7b3de7354/src/Pulp/Server.purs#L99

If that line is removed I think things will work.

**Comment by pkamenarsky** [comment]
Btw, how can I use your sample webpack config? I placed it in `webpack.config.js` and invoked `pulp server`, same issue.

**Comment by hdgarrood** [comment]
From the output of `pulp server --help`:

```
  --config -c <file>          Override the default Webpack config.
```

So `pulp server --config webpack.config.js`.

**Comment by pkamenarsky** [comment]
Allright, now I have this webpack config: https://gist.github.com/pkamenarsky/bd9f262841794de17b01

`purs-loader: 0.5.0`
`purescript-webpack-plugin: 0.1.1`

`src\webpack.js` looks like this:

    // module Webpack

    require('./Main').main();

Then I do:

    pulp server -d --config webpack.config.js

Still:

    WARNING in ./output/react/foreign.js
    There is another module with an equal name when case is ignored.
    This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
    Rename module if multiple modules are expected or use equal casing if one module is expected.

**Comment by ethul** [comment]
In `output`, do you have `output/react` and `output/React`?

**Comment by ethul** [comment]
Also, in `bower_components`, do you have `bower_components/react`?

**Comment by pkamenarsky** [comment]
Only `React` in `output` and no `react` in `bower_components`.

**Comment by ethul** [comment]
I am curious why this is closed. Was there a fix?

**Comment by pkamenarsky** [comment]
I'm not sure - I managed to get it to work using the webpack config from `purs-loader`, but maybe this is a bug in `pulp`?

**Comment by hdgarrood** [comment]
Yes, this is a bug in `pulp`, which could be fixed by addressing bodil/pulp#151.

**Comment by ethul** [comment]
@pkamenarsky and @hdgarrood Thanks for helping out and digging into this issue. I am glad it ended up working.

Regarding bodil/pulp#151, if it turns out that the plan is to not swap out webpack, I'd be happy to work on a PR to update to the newer purs-loader/purescript-webpack-plugin, as well as update the default webpack.config.js. But let me know what is decided.
